### DISCUSSION TITLE
Customizing InputArgument parsing

### DISCUSSION THREAD

**Discussion by beikov** [Netflix/dgs-framework#775] [Q&A]
Hi,

I saw that the input argument handling was made completely static with this change: https://github.com/Netflix/dgs-framework/commit/2fe40a9fd1acab3037354586338dcede7c4ea9b7#diff-0d8ecc929924d083759c48f18c0bda7a0671208191ccdd30ac743aaedd264b4e

and before it was only possible to "hack" into `DgsSchemaProvider` to change the internal `ObjectMapper` for custom deserialization logic.

Can you please provide a hook for replacing the parsing of InputArgument?

My main use case is that I want to accept an interface as input argument for a mutation. Apart from that, my desired deserializer also controls how objects are constructed.

**Comment by paulbakker** [comment]
This issue contains the reasoning for the custom implementation and not using Jackson: https://github.com/Netflix/dgs-framework/pull/549

In short, the input we're getting is not JSON, so handling it as JSON _seems_ to work in many scenarios, but its fundamentally broken. This means that replacing the implementation would be far more involved than just using something like Jackson, and will most likely introduce other issues. Because it's such as hairy topic, I'm thinking it's probably not a good idea to make this customizable, because it's going to cause more issues than it solves.

What we could consider is adding support for the specific use cases you find missing in the implementation.

**Reply by beikov** [reply]
I don't care how the string is parsed. In fact, I don't even want to know about the parsing :) I can imagine though that there are some parsing requirements with respect to support various basic types e.g. timestamp etc., but that's a different topic.
All I need is control over how objects are constructed and how attributes are set on the objects, so if you could introduce an abstraction for that, this would be enough for my purposes. Something as simple as this for example:

```
interface ObjectMapper {
  <T> T fromAttributes(Map<String, Object> map, java.lang.reflect.Type t);
  default <T> T fromAttributes(Map<String, Object> map, Class<T> clazz) {
    return (T) formAttributes(map, (Type) clazz);
  }
}
```

Where the value for a map entry referring to a nested object would be a `Map<String, Object>` again.

**Comment by paulbakker** [comment]
Isn't that the same as using `DataFetchingEnvironment.getArgument(...)`
directly?

**Reply by beikov** [reply]
I want users to be able to write code like this:

```
@DgsMutation
public Long createCat(@InputArgument(name = "cat") CatCreateDto dto) {
...
}
```

and for this to work, a contract is needed that handles the construction of argument objects. I don't want to force users to bind to some implementation for the parsing/construction of the object. That's just boilerplate that the declarative model of DGS should get rid off, no?

**Comment by EugenMayer** [comment]
@beikov hey :) Just to understand the issue a little better and be part of the discussion, could you make it a little more clear if this is a conveniens effort to make the input argument binding easier with BP is used, or is this a critical requirement to get mutations going with blaze persistence?

@paulbakker as a side note, we are heavy users of blaze-persistence which @beikov is the author of, and we worked with him in the last months to get a DGS integration up and running, which we are currently already using in production (you might know, we had a couple of discussions/bugs we looked at with you together). This said we are yet only using fetchers, mutations are on our list. Regarding how deeply integrated BP is in our persistence model it would be a huge bummer if mutations are not working with DGS - this would mean we would need to remove DGS and replace it with something else (which we actually do not want to). We would love to use DGS and BP together, it is actually a great match!

**Reply by beikov** [reply]
DGS parses the input based on the method parameter types and passes the parsed object to the method, so I would say this is critical for mutations to work nicely with any library/app that requires custom deserialization.

Since DGS doesn't support inferring the schema and requires to define that separately, there is of course the option of burdening the user with writing the parsing boilerplate code over and over again, like this:

```java
@Autowired
ObjectMapper customObjectMapper;

@DgsMutation
public Long createCat(@InputArgument(name = "cat") String payload) {
    CatCreateDto cat = customObjectMapper.fromJson(payload, CatCreateDto.class);
    ...
}
```

**Reply by EugenMayer** [reply]
Due to the lack of hands-on with mutations, excuse my lack of detail here.

I guess you are referring to https://netflix.github.io/dgs/mutations and thus AFAICS in DGS, mutations can be used using

1. `getArgument` to fetch single fields (simple types) and then use them the manual way
2. `getArgument` to fetch a map and the use the `ObjectMapper` to map those on models
3. `InputArgument` to pick a filed from the request and map it on a `model` directly (in the signature) using the default ObjectMapper
 
So for BP as I know, we need to use a custom deserializer so the automatic EM/EVM magic can happen so the EntityViews (Create/Update views) can be properly created.

I guess we can implement this with DGS using 2 ways:

a) being able to replace the ObjectMapper for deserialization as you suggested here, very similar how deserialization is pluggable in spring MVC and already integrated with BP
b) offer a tool-method, let's call it `T model = EVMGQ.mapModel(DataFetchingEnvironment dfe, String inputName)` or similar 

While a) requires that DGS offers the ability to make the ObjectMapper replaceable like SpringMVC I guess b) does not. 
That is about right, @beikov?

---

Of course I would be very happy to have a) so it all aligns with other `Spring Boot` technologies and the same way that the `ObjectMapper` is integrated in SpringMVC - having the same integration point and interface would not only make a lot of sense - it would also make the integration a lot more convenient and less error prone. calling b) would required specific logic and does not abstract away from this.

This said, i'am not sure what specific @paulbakker was talking about when judging that making the ObjectMapper interchangeable would possible open it up more then DGS expects or can offer. Could you go into a little more detail here Paul? Maybe there is a middle ground?

**Reply by paulbakker** [reply]
**First a bit more context.**
One important aspect to understand is that DGS is *not* using an `ObjectMapper`. `ObjectMapper` is specifically from the Jackson library, and is for JSON (de)serialization. The Jackson ObjectMapper has many ways it can be configured, hence it being plug-able in Spring. It's only for JSON though.

In DGS we're not using the `ObjectMapper` (for input arguments), because the input is not JSON. In an older version of the framework we _did_ use it, specifically the `convert` method, which takes a Map, serializes it to a JSON String, and then deserializes it back to an object. However, that didn't work well, because that completely ignores scalars. This was a hacky implementation that caused all sorts of problems related to scalars, so we had to replace that mechanism.

The implementation we now use is conceptually similar to `ObjectMapper`, but technically different. It's based on the fact that `graphql-java` gives us a `Map` with already converted scalar values. Our code takes that map, and maps it onto a Java object, so that the user can specifcy something like `@InputArgument Cat cat`.
The code that does this is [here](https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/InputObjectMapper.kt). As you can see in the implementation, it's quite complex because there are a lot of corner cases to account for.

**Ideas**
Because we're not using `ObjectMapper`, there isn't anywhere to plug-in your own instance of it, so that's not an option.
What we can discuss however, is another hook that you could use to do the conversion you need. One way would be to make our [InputObjectMapper](https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/InputObjectMapper.kt) replace-able. That's probably not great though, because re-implementing that code is probably not what you want to do, since it's so complex.

What would be the most useful hook for you to have?

**Reply by EugenMayer** [reply]
Thank you a lot for those details. The bottom line is, `ObjectMapper` is no longer used starting from 4.6 since it was basically missed beforehand (to json->from json).

The new implementation fixes that unneeded toJson conversion, but is a new implementation and thus the integration @beikov would like to reuse does not work out of the box - there is no integration layer right now.

Creating a new integration layer:
Re-Implementing the mapping should not re-implement the entire InputMapperObject, but i'am not sure this can be avoided. Thus we need to define what we will need for BP to map into it's own models.

@beikov AFAIU the actual "middle ground" is a map with scalars.

We could go either with

a) Reusing that map and try to not re-implement `InputObjectMapper`, this would need adjustments to the current deserialization code and i'am not sure how easy that would be, but i assume that could work out fairly well

b) Being able to replace the InputObjectMapper using a interface that is introduced by DGS. If i'am not wrong, the current implementation looks similar to what we already have in BP, the only question is if `mapToKotlinObject` needs to be supported by our implementation (@paulbakker why is that needed, compared to `mapToJavaObject`?)

Any thoughts on that @beikov ?

**Reply by beikov** [reply]
So I think some kind of contract that allows delegating to the default implementation for sub-graphs would be best.

For updatable/createble entity views we have 2 concerns that are important for us to be able to control

* Construction of the object and being able to "hide"/"remove" an element from the map
* Use `Collection#add`/`Map#put` to deserialize contents onto `Collection`/`Map` model types instead of constructing whole collections and apply them via a setter

The first part is important because some entity views have a special constructor, accepting the identifier of underlying row which must be consumed in a special way as that is important for deciding whether to use update/create semantics.

The second part is important because updatable/creatable entity views use special collection implementations that record collection operations and shouldn't be replaced.

Maybe an SPI like this would work:

```
interface InputArgumentMapper {
    default Object mapToJavaObject(Map<String, Object> inputMap, java.lang.reflect.Type targetType, InputArgumentMapper defaultMapper) {
        return defaultMapper.mapToJavaObject(inputMap, targetType, defaultMapper);
    }
    default Object mapToKotlinObject(Map<String, Object> inputMap, KType targetType, InputArgumentMapper defaultMapper) {
        return defaultMapper.mapToKotlinObject(inputMap, targetType, defaultMapper);
    }
}
```

**Reply by paulbakker** [reply]
I have a PR ready with the proposed SPI: https://github.com/Netflix/dgs-framework/pull/842.
@beikov @EugenMayer please review if this works for you.

Example use can be found in the tests provided in the PR.

**Reply by beikov** [reply]
May I ask how arrays in the payload are deserialized in the map?

**Reply by paulbakker** [reply]
A value in the map of inputs can be of type `List`. For an array of objects, the items in the list would be another `Map<String, Object>`. For "simple" values, they are the converted scalar values.

For further context, this is the code that converts Lists in our InputObjectMapper: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/InputObjectMapper.kt#L171
Unfortunately its quite complex due to generics.
Do you need another method to hook into in interface I added for lists?

**Reply by beikov** [reply]
Well, I was just thinking about how I will need to construct the whole object graph now from within that single method and was curious about collections, but it sounds to me like this should work out. My original suggestion to use `java.lang.reflect.Type` and `KType` could be used to implement a recursive approach per field, but I guess this works good enough as well.

**Reply by EugenMayer** [reply]
Just that you guys know, i still follow, but i cannot really add much - @beikov does all the work here and has about 1 gazillion more insight, so i just keep quiet and follow.

Super happy though that this issue is moving forward, thank you so much for the effort!

**Reply by paulbakker** [reply]
@beikov Also, if this wasn't clear yet, for each level of object in the graph your custom mapper will be called again: https://github.com/Netflix/dgs-framework/pull/842/files#diff-674a7494b4ba6668d6f6e9af0b0ad17bc3c672e13f0fdc65cee6a62a94998138R81.

The tests in the PR demonstrate this as well. This means you don't have to construct the whole graph from the root, you can make modifications at each child object and call back into the default mapper.

**Reply by beikov** [reply]
Yeah, I guess this should work for me. Thanks!

**Comment by EugenMayer** [comment]
@beikov @paulbakker we are now at at stage were we finally start with mutators and now hit this thing.

I did not remember and find this issue directly, thus search around custom scalars and things like that, leading to nothing when going down the road.

As far as i can see @beikov this is a similar take then we alware have with the spring-webmvc "jackson" integration (`EntityViewAwareMappingJackson2HttpMessageConverter) which works as an converter registered with `BlazePersistenceWebConfiguration` in the WebMVC framework.

Currently, seeing the discussion around `InputObjectMapper` which do around the  same job as `EntityViewAwareMappingJackson2HttpMessageConverter` i ask myself:

- How is an customize the InputObjectMapper implementation - i see there is `DefaultInputObjectMapper` and is created as a bean via `DgsInputArgumentConfiguration` - so is the intention to to decorate / inherit DefaultInputObjectMapper, expose the new `DefaultInputObjectMapper` as `primary` and the add the needed handling for custom / other input types?
- the discussion aged a bit and we are not looking at 5.x / 6.x - did anything change in the way this can be done that needs a refresher?

Thank you for the effort!

**Reply by beikov** [reply]
The idea is simple, an integration in Blaze-Persistence provides a bean for the `InputObjectMapper` contract which is then used to implement deserialization based on entity view logic, and falls back to the `DefaultInputObjectMapper` for non entity view types.
